File: scripts/validate_atlas.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DEFAULT = ROOT / "data" / "graph" / "cannabiverse_graph_v0.4.json"
GRAPH = Path(sys.argv[1]) if len(sys.argv) > 1 else DEFAULT

errors: list[str] = []
warnings: list[str] = []

def fail(msg: str) -> None:
    errors.append(msg)

def warn(msg: str) -> None:
    warnings.append(msg)

def require(cond: bool, msg: str) -> None:
    if not cond:
        fail(msg)

if not GRAPH.exists():
    fail(f"graph file not found: {GRAPH}")
else:
    try:
        data = json.loads(GRAPH.read_text(encoding="utf-8"))
    except Exception as exc:
        fail(f"cannot parse JSON: {exc}")
        data = {}

if not errors:
    compounds = data.get("compounds", [])
    aliases = data.get("aliases", [])
    edges = data.get("relationships", [])
    sources = data.get("sources", [])
    process_nodes = data.get("process_nodes", [])
    provisional = data.get("provisional_candidates", [])
    meta = data.get("metadata", {})

    compound_ids = [c.get("atlas_id") for c in compounds]
    source_ids = [s.get("source_id") for s in sources]
    process_ids = [p.get("node_id") for p in process_nodes]
    provisional_ids = [p.get("candidate_id") for p in provisional]

    require(len(compound_ids) == len(set(compound_ids)), "duplicate canonical atlas_id detected")
    require(len(source_ids) == len(set(source_ids)), "duplicate source_id detected")
    require(len(process_ids) == len(set(process_ids)), "duplicate process node_id detected")
    require(len(provisional_ids) == len(set(provisional_ids)), "duplicate provisional candidate_id detected")

    for cid in compound_ids:
        require(bool(cid and re.fullmatch(r"CPA-CAN-\d{4}", cid)), f"invalid canonical ID: {cid!r}")

    baseline = [c for c in compounds if c.get("canonical_status") == "BASELINE_2021_LOCKED"]
    require(len(baseline) == 125, f"baseline must remain exactly 125 records, found {len(baseline)}")
    baseline_ids = {c.get("atlas_id") for c in baseline}
    expected_baseline_ids = {f"CPA-CAN-{i:04d}" for i in range(1, 126)}
    require(baseline_ids == expected_baseline_ids, "baseline permanent ID set changed; expected CPA-CAN-0001..0125")

    require(meta.get("canonical_compound_count") == len(compounds), "metadata canonical_compound_count does not match compounds[]")
    require(meta.get("historical_baseline_count") == len(baseline), "metadata historical_baseline_count mismatch")
    require(meta.get("post_baseline_confirmed_count") == len(compounds) - len(baseline), "metadata post_baseline_confirmed_count mismatch")
    require(meta.get("provisional_candidate_count") == len(provisional), "metadata provisional_candidate_count mismatch")
    require(meta.get("relationship_edge_count") == len(edges), "metadata relationship_edge_count mismatch")

    source_set = set(source_ids)
    canonical_set = set(compound_ids)
    process_set = set(process_ids)
    provisional_set = set(provisional_ids)
    all_nodes = canonical_set | process_set | provisional_set

    require(not (canonical_set & provisional_set), "a provisional candidate ID also appears in canonical compounds")

    for c in compounds:
        cid = c.get("atlas_id")
        require(bool(c.get("canonical_name")), f"{cid}: missing canonical_name")
        require(c.get("source_id") in source_set, f"{cid}: unresolved source_id {c.get('source_id')!r}")
        require(bool(c.get("source_url")), f"{cid}: missing source_url")
        require(bool(re.match(r"^I[0-4](?:_|$)", str(c.get("identity_level", "")))), f"{cid}: invalid identity_level {c.get('identity_level')!r}")
        require(bool(re.match(r"^O[0-4](?:_|$)", str(c.get("occurrence_level", "")))), f"{cid}: invalid occurrence_level {c.get('occurrence_level')!r}")
        require(bool(re.match(r"^B[0-3](?:_|$)", str(c.get("biosynthesis_level", "")))), f"{cid}: invalid biosynthesis_level {c.get('biosynthesis_level')!r}")
        require(c.get("canonical_status") in {"BASELINE_2021_LOCKED", "POST_BASELINE_CONFIRMED"}, f"{cid}: invalid canonical_status")

    for a in aliases:
        require(a.get("atlas_id") in canonical_set, f"alias points to missing canonical compound: {a}")
        require(bool(a.get("alias")), f"empty alias: {a}")

    for p in provisional:
        pid = p.get("candidate_id")
        require(bool(pid and re.fullmatch(r"CPA-PROV-\d{4}", pid)), f"invalid provisional ID: {pid!r}")
        require(p.get("source_id") in source_set, f"{pid}: unresolved source_id {p.get('source_id')!r}")
        require(bool(re.match(r"^I[0-4](?:_|$)", str(p.get("identity_level", "")))), f"{pid}: invalid identity_level")
        require(bool(re.match(r"^O[0-4](?:_|$)", str(p.get("occurrence_level", "")))), f"{pid}: invalid occurrence_level")

    edge_ids = [e.get("edge_id") for e in edges]
    require(len(edge_ids) == len(set(edge_ids)), "duplicate edge_id detected")
    for e in edges:
        eid = e.get("edge_id")
        require(bool(eid and re.fullmatch(r"CPA-EDGE-\d{4}", eid)), f"invalid edge ID: {eid!r}")
        require(e.get("source_node") in all_nodes, f"{eid}: unresolved source_node {e.get('source_node')!r}")
        require(e.get("target_node") in all_nodes, f"{eid}: unresolved target_node {e.get('target_node')!r}")
        require(bool(e.get("relation")), f"{eid}: missing relation")
        require(e.get("source_id") in source_set, f"{eid}: unresolved source_id {e.get('source_id')!r}")
        require(bool(e.get("source_url")), f"{eid}: missing source_url")
        lane = str(e.get("evidence_lane", ""))
        conf = str(e.get("confidence", ""))
        if "hypothesis" in lane.lower():
            require(conf.startswith("HYPOTHESIS"), f"{eid}: hypothesis lane must use hypothesis confidence, found {conf!r}")

    for e in edges:
        if e.get("source_node") in provisional_set or e.get("target_node") in provisional_set:
            warn(f"{e.get('edge_id')}: relationship touches provisional candidate; review authority manually")

    for r in data.get("reconciliation_log", []):
        require(r.get("source_id") in source_set, f"reconciliation event {r.get('event_id')}: unresolved source_id")

    gov = str(meta.get("governance_rule", ""))
    require("IDENTITY" in gov and "OCCURRENCE" in gov, "metadata governance_rule missing lane-separation doctrine")

print(f"Cannabiverse validator: {GRAPH}")
for w in warnings:
    print(f"WARN: {w}")
if errors:
    for e in errors:
        print(f"ERROR: {e}")
    print(f"FAIL: {len(errors)} error(s), {len(warnings)} warning(s)")
    raise SystemExit(1)

print(
    f"PASS: {len(data.get('compounds', []))} canonical compounds, "
    f"{len(data.get('relationships', []))} edges, "
    f"{len(data.get('sources', []))} sources, "
    f"{len(warnings)} warning(s)"
)

File: scripts/validate-atlas.mjs
import fs from "node:fs";
import path from "node:path";

const root = process.cwd();
const readJson = p => JSON.parse(fs.readFileSync(path.join(root, p), "utf8"));
const fail = msg => { throw new Error(msg); };
const pass = msg => console.log(`PASS  ${msg}`);

const graph = readJson("data/graph/cannabiverse_graph_v0.4.json");
const compounds = graph.compounds ?? [];
const aliases = graph.aliases ?? [];
const edges = graph.relationships ?? [];
const sources = graph.sources ?? [];
const provisional = graph.provisional_candidates ?? [];
const processNodes = graph.process_nodes ?? [];

const claims = readJson("data/evidence/relationship_claims.json");
const sourceNodes = readJson("data/evidence/source_nodes.json");
const supportEdges = readJson("data/evidence/source_claim_edges.json");
const sourceSummary = readJson("data/evidence/source_claim_summary.json");

const sourceIds = new Set(sources.map(s => s.source_id));
const compoundIds = new Set(compounds.map(c => c.atlas_id));
const processIds = new Set(processNodes.map(n => n.node_id));
const claimIds = new Set(claims.map(c => c.claim_id));

if (compoundIds.size !== compounds.length) fail("duplicate canonical atlas_id");
pass("all canonical Atlas IDs are unique");

if (!compounds.every(c => /^CPA-CAN-\d{4}$/.test(c.atlas_id))) fail("invalid canonical ID format");
pass("canonical IDs use CPA-CAN-#### format");

const baseline = compounds.filter(c => c.canonical_status === "BASELINE_2021_LOCKED");
if (baseline.length !== 125) fail(`baseline count changed: ${baseline.length}`);
const baselineIds = baseline.map(c => Number(c.baseline_id)).sort((a,b)=>a-b);
if (baselineIds.some((v,i)=>v !== i+1)) fail("baseline IDs are not exactly 1..125");
pass("Radwan-125 historical baseline is complete and immutable in count/numbering");

if (compounds.length !== graph.metadata.canonical_compound_count) fail("metadata canonical count mismatch");
if (edges.length !== graph.metadata.relationship_edge_count) fail("metadata edge count mismatch");
if (provisional.length !== graph.metadata.provisional_candidate_count) fail("metadata provisional count mismatch");
pass("metadata counts agree with graph arrays");

for (const c of compounds) {
  if (!sourceIds.has(c.source_id)) fail(`compound ${c.atlas_id} references missing source ${c.source_id}`);
  if (!/^I[0-4](?:_|$)/.test(c.identity_level)) fail(`bad identity level on ${c.atlas_id}: ${c.identity_level}`);
  if (!/^O[0-4](?:_|$)/.test(c.occurrence_level)) fail(`bad occurrence level on ${c.atlas_id}: ${c.occurrence_level}`);
  if (!/^B[0-3](?:_|$)/.test(c.biosynthesis_level)) fail(`bad biosynthesis level on ${c.atlas_id}: ${c.biosynthesis_level}`);
}
pass("canonical provenance and evidence vocabularies resolve");

for (const a of aliases) {
  if (!compoundIds.has(a.atlas_id)) fail(`alias points to missing compound ${a.atlas_id}`);
}
pass("all aliases resolve to canonical compounds");

for (const p of provisional) {
  if (!/^CPA-PROV-\d{4}$/.test(p.candidate_id)) fail(`bad provisional ID ${p.candidate_id}`);
  if (compoundIds.has(p.candidate_id)) fail(`provisional ID overlaps canon ${p.candidate_id}`);
  if (!sourceIds.has(p.source_id)) fail(`provisional ${p.candidate_id} references missing source`);
}
pass("provisional namespace is separate and source-resolved");

const allowedLanes = new Set(["biosynthesis","transformation","structure","biosynthesis_hypothesis"]);
for (const e of edges) {
  if (!allowedLanes.has(e.evidence_lane)) fail(`unknown edge lane ${e.edge_id}: ${e.evidence_lane}`);
  const sourceNodeOkay = compoundIds.has(e.source_node) || processIds.has(e.source_node);
  const targetNodeOkay = compoundIds.has(e.target_node) || processIds.has(e.target_node);
  if (!sourceNodeOkay) fail(`edge ${e.edge_id} missing source node ${e.source_node}`);
  if (!targetNodeOkay) fail(`edge ${e.edge_id} missing target node ${e.target_node}`);
  if (!sourceIds.has(e.source_id)) fail(`edge ${e.edge_id} references missing scientific source ${e.source_id}`);

  const hypothesisSignal =
    e.evidence_lane.includes("hypothesis") ||
    String(e.confidence).includes("HYPOTHESIS") ||
    String(e.relation).startsWith("POSSIBLE_") ||
    String(e.relation).startsWith("PROPOSED_");

  if (hypothesisSignal && e.evidence_lane !== "biosynthesis_hypothesis") {
    fail(`hypothesis edge ${e.edge_id} is not kept in biosynthesis_hypothesis lane`);
  }
}
pass("all graph endpoints and edge sources resolve");
pass("hypothesis relationships remain explicitly typed as hypotheses");

// v0.2 claim/source projection checks.
if (claims.length !== edges.length) fail(`claim/edge cardinality mismatch: ${claims.length} claims vs ${edges.length} edges`);
if (claimIds.size !== claims.length) fail("duplicate claim IDs");

const edgeById = new Map(edges.map(e => [e.edge_id, e]));
for (const c of claims) {
  if (!/^CPA-CLM-\d{4}$/.test(c.claim_id)) fail(`invalid claim ID ${c.claim_id}`);
  const e = edgeById.get(c.derived_from_edge_id);
  if (!e) fail(`claim ${c.claim_id} references missing edge ${c.derived_from_edge_id}`);
  if (c.subject_id !== e.source_node) fail(`claim ${c.claim_id} subject drift`);
  if (c.predicate !== e.relation) fail(`claim ${c.claim_id} predicate drift`);
  if (c.object_id !== e.target_node) fail(`claim ${c.claim_id} object drift`);
  if (c.evidence_lane !== e.evidence_lane) fail(`claim ${c.claim_id} lane drift`);
  if (c.authority !== e.confidence) fail(`claim ${c.claim_id} authority drift`);
  if (c.source_ids.length !== 1 || c.source_ids[0] !== e.source_id) fail(`claim ${c.claim_id} source drift`);
  const expectedStatus = e.evidence_lane === "biosynthesis_hypothesis" ? "HYPOTHESIS" : "ASSERTED_WITH_SOURCE";
  if (c.status !== expectedStatus) fail(`claim ${c.claim_id} status drift`);
}
pass("every chemistry edge has an exact deterministic claim receipt");

if (sourceNodes.length !== sources.length) fail("source-node/source-register cardinality mismatch");
for (const n of sourceNodes) {
  if (!sourceIds.has(n.node_id)) fail(`source node ${n.node_id} does not resolve`);
}
pass("all registered scientific sources are first-class source nodes");

if (supportEdges.length !== claims.length) fail("source-support edge count must equal relationship-claim count");
for (const e of supportEdges) {
  if (!sourceIds.has(e.source_node)) fail(`support edge has missing source ${e.source_node}`);
  if (!claimIds.has(e.target_node)) fail(`support edge has missing claim ${e.target_node}`);
  if (e.relation !== "SUPPORTS_CLAIM" || e.evidence_lane !== "source_support") fail(`invalid support-edge semantics ${e.edge_id}`);
}
pass("source → claim provenance graph is complete");

if (sourceSummary.length !== sources.length) fail("source summary does not cover all sources");
const summarizedClaims = sourceSummary.reduce((n,s)=>n + Number(s.relationship_claim_count||0),0);
if (summarizedClaims !== claims.length) fail("source summary claim total mismatch");
pass("source claim summary reconciles to the claim ledger");

// Verify the diff-friendly split canon reconstructs the embedded full graph.
const canonDir = path.join(root, "data", "canon");
const compoundFiles = fs.readdirSync(canonDir).filter(n => /^compounds-\d{3}-\d{3}\.json$/.test(n)).sort();
const aliasFiles = fs.readdirSync(canonDir).filter(n => /^aliases-\d{3}-\d{3}\.json$/.test(n)).sort();
const splitCompounds = compoundFiles.flatMap(n => readJson(path.join("data","canon",n)));
const splitAliases = aliasFiles.flatMap(n => readJson(path.join("data","canon",n)));

if (JSON.stringify(splitCompounds) !== JSON.stringify(compounds)) fail("split compound canon does not reproduce graph compounds");
if (JSON.stringify(splitAliases) !== JSON.stringify(aliases)) fail("split aliases do not reproduce graph aliases");
pass("diff-friendly canon chunks reproduce the full graph exactly");

console.log("");
console.log(`VALID  Cannabiverse: ${compounds.length} compounds, ${aliases.length} aliases, ${edges.length} chemistry edges, ${claims.length} claim receipts, ${sources.length} source nodes, ${provisional.length} provisional/search-target records.`);
